File: src/main/scala/com/github/wenjunhuang/lox/Statement.scala
package com.github.wenjunhuang.lox

enum Statement:
  case Expr(expression: Expression)
  case Print(expression: Expression)
  case Var(name: Token, initializer: Option[Expression])
  case Block(statements: Vector[Statement])

  def accept[T](visitor: StatementVisitor[T]): T = this match
    case v: Var       => visitor.visitVarStatement(v)
    case expr: Expr   => visitor.visitExpressionStatement(expr)
    case print: Print => visitor.visitPrintStatement(print)
    case block: Block => visitor.visitBlockStatement(block)

end Statement

File: fsharp/library/Statement.fs
namespace com.github.wenjunhuang.lox

type Statement =
    | Expr of expr: Expression
    | Var of name: Token * initializer: Expression option
    | If of condition: Expression * thenBranch: Statement * elseBranch: Statement option
    | Print of expr: Expression
    | Block of statements: array<Statement>
